apps/src/blockly/addons/cdoUtils.ts

```diff
@@ -58,6 +59,48 @@ export function loadBlocksToWorkspace(
   Blockly.serialization.workspaces.load(mainSource, workspace);
   positionBlocksOnWorkspace(workspace);
   Blockly.hasLoadedBlocks = true;
+
+  // Dynamically add procedure call blocks to an uncategorized toolbox
+  // if specified in the level config (e.g. Minecraft Agent levels).
+  // Levels will include: "top_level_procedure_autopopulate": "true"
+  if (Blockly.topLevelProcedureAutopopulate) {
+    addProcedureCallBlocksToFlyout(workspace, mainSource);
+  }
+}
+
+function addProcedureCallBlocksToFlyout(
+  workspace: WorkspaceSvg,
+  mainSource: WorkspaceSerialization
+) {
+  // options.languageTree is the translated toolbox info
+  if (workspace.getFlyout() && workspace.options?.languageTree) {
+    const callBlocks = [] as ToolboxItemInfo[];
+    const defBlocks = mainSource.blocks.blocks.filter(
+      block => block.type === 'procedures_defnoreturn'
+    );
+    defBlocks.forEach(def => {
+      // Procedure definitions should have a valid name
+      if (typeof def.fields?.NAME === 'string') {
+        // Create the block XML for a procedure call block.
+        const block = document.createElement('block');
+        block.setAttribute('type', 'procedures_callnoreturn');
+        const mutation = document.createElement('mutation');
+        mutation.setAttribute('name', def.fields.NAME);
+        block.appendChild(mutation);
+
+        callBlocks.push({
+          kind: 'BLOCK',
+          blockxml: block,
+          type: 'procedures_callnoreturn',
+        });
+      }
+    });
+    if (callBlocks.length) {
+      // Add the new callblocks to the toolbox and refresh it.
+      workspace.options.languageTree.contents.push(...callBlocks);
```

Comment:
Is this the first place we're using language tree like this? It feels new to me, but I'm surprised to feel like I'm seeing it for the first time.

Reply:
See updated description and let me know if you have more questions. It's new to the main repo, but supported in both Blockly versions.